rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java

```diff
@@ -75,8 +99,202 @@ public NodesLockRestorationManager apply(RMCore rmCore) {
         };
     }
 
-    public void restoreLocks(RMNode rmNode, Client caller) {
-        nodesLockRestorationManager.handle(rmNode, caller);
+    public void recoverNodeSourcesAndNodes() {
+        Collection<NodeSourceData> nodeSources = rmCore.getDbManager().getNodeSources();
+        logPersistedNodeSourceInfo(nodeSources);
+
+        for (NodeSourceData nodeSourceData : nodeSources) {
+            String nodeSourceName = nodeSourceData.getName();
+            if (NodeSource.DEFAULT_LOCAL_NODES_NODE_SOURCE_NAME.equals(nodeSourceName)) {
+                // will be recreated by SchedulerStarter
+                rmCore.getDbManager().removeNodeSource(nodeSourceName);
+            } else {
+                recoverNodeSourceSuccessfullyOrRemove(nodeSourceData, nodeSourceName);
+            }
+        }
+    }
+
+    public boolean recoverFullyDeployedInfrastructureOrReset(String nodeSourceName, NodeSource nodeSourceToDeploy,
+            NodeSourceDescriptor descriptor) {
+        boolean recoverNodes = false;
+        boolean existPersistedNodes = existPersistedNodes(nodeSourceName);
+
+        if (existPersistedNodes) {
+            InfrastructureManager im = InfrastructureManagerFactory.recover(NodeSourceData.fromNodeSourceDescriptor(descriptor));
+            if (!im.getDeployingAndLostNodes().isEmpty()) {
+                // if there are deploying nodes, we will not recover
+                rmCore.getDbManager().removeAllNodesFromNodeSource(nodeSourceName);
+            } else {
+                recoverNodes = true;
+                nodeSourceToDeploy.setInfrastructureManager(im);
+            }
+        }
+        return recoverNodes;
+    }
+
+    private boolean existPersistedNodes(String nodeSourceName) {
+        boolean existPersistedNodes = false;
+
+        Collection<RMNodeData> nodesData = rmCore.getDbManager().getNodesByNodeSource(nodeSourceName);
+        if (nodesData.isEmpty()) {
+            logger.info("No node found in database for node source: " + nodeSourceName);
+        } else {
+            existPersistedNodes = true;
+        }
+
+        return existPersistedNodes;
+    }
+
+    private void recoverNodeSourceSuccessfullyOrRemove(NodeSourceData nodeSourceData, String nodeSourceName) {
+        try {
+            logger.info("Recover node source " + nodeSourceName);
+            // retrieve node source status
+            boolean deployNodeSource = false;
+            if (nodeSourceData.getStatus().equals(NodeSourceStatus.NODES_DEPLOYED)) {
+                // reset node source status to be able to deploy again
+                nodeSourceData.setStatus(NodeSourceStatus.NODES_UNDEPLOYED);
+                deployNodeSource = true;
+            }
+            rmCore.prepareNodeSource(nodeSourceData);
+            if (deployNodeSource) {
+                rmCore.deployNodeSource(nodeSourceName);
+            }
+        } catch (Throwable t) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Catch Exception instead of Throwable. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1181)